crates/typst-cli/src/watch.rs

```diff
@@ -34,20 +34,45 @@ pub fn watch(mut command: CompileCommand) -> StrResult<()> {
     let timeout = std::time::Duration::from_millis(100);
     let output = command.output();
     loop {
+        let mut removed = HashSet::new();
         let mut recompile = false;
         for event in rx
             .recv()
             .into_iter()
             .chain(std::iter::from_fn(|| rx.recv_timeout(timeout).ok()))
         {
             let event = event.map_err(|_| "failed to watch directory")?;
+
+            // Workaround for notify-rs' implicit unwatch on remove/rename (triggered
+            // by some editors when saving files) with the inotify backend.
+            // By keeping track of the removed files, we can allow those we still
+            // depend on to be watched again later on.
+            if matches!(
+                event.kind,
+                notify::EventKind::Remove(notify::event::RemoveKind::File)
+            ) {
+                let path = &event.paths[0];
+                removed.insert(path.clone());
+
+                // Remove the watch if it still exists.
+                if let Err(notify::Error { kind, .. }) = watcher.unwatch(path) {
+                    match kind {
+                        notify::ErrorKind::WatchNotFound => {}
+                        _ => return Err("failed to unwatch file".into()),
+                    }
+                }
```

Comment:
Do we really want this to return for `kind != notify::ErrorKind::WatchNotFound`? In the `watch_dependencies` function below I just did `watcher.unwatch(&path).ok()` to ignore the result. After all, the unwatching is only an optimization basically.